test/contracts/HCElection.aes

```diff
@@ -245,6 +240,28 @@ main contract HCElection =
   function pay_reward_((acc, tot), (addr, amt)) =
     (acc{[addr = 0] @ r = r + amt}, tot + amt)
 
+  stateful function collect_penalties(epoch : int) =
+    let ei = state.epochs[epoch]
+    let pens = collect_penalties_(ei.start, ei.length, {})
+    state.main_staking_ct.add_penalties(epoch, List.sort(compare_pens_, pens)) // penalties first to fill pool, rewards second
+
+  stateful function
+    collect_penalties_(_, 0, acc) = Map.to_list(acc)
+    collect_penalties_(h, n, acc) =
+      switch(Map.lookup(h, state.penalties))
+        None      => collect_penalties_(h + 1, n - 1, acc)
+        Some(aas) =>
+          let acc1 = List.foldl(collect_penalty_, acc, aas)
+          put(state{penalties @ r = Map.delete(h, r)})
+          collect_penalties_(h + 1, n - 1, acc1)
```

Comment:
This is not super-nice, this is almost the same function again (`pay_rewards`) - but let's ignore that.